File: crates/domain/src/library/scan_state.rs
use jiff::Timestamp;

use crate::library::LibraryId;

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub enum ScanStatus {
    Idle,
    Queued,
    Running,
    Failed,
}

#[derive(Debug, Clone)]
pub struct ScanState {
    pub library: LibraryId,
    pub status: ScanStatus,
    pub progress: f32,
    pub started_at: Option<Timestamp>,
    pub last_scanned_at: Option<Timestamp>,
    pub error: Option<String>,
}

File: clients/flutter/lib/model/library/scan_state.dart
import 'package:freezed_annotation/freezed_annotation.dart';

part 'scan_state.freezed.dart';
part 'scan_state.g.dart';

enum ScanStatus {
  @JsonValue('idle')
  idle,
  @JsonValue('queued')
  queued,
  @JsonValue('running')
  running,
  @JsonValue('failed')
  failed,
}

@freezed
abstract class ScanState with _$ScanState {
  const factory ScanState({
    required String libraryId,
    required ScanStatus status,
    @Default(0) double progress,
    String? startedAt,
    String? lastScannedAt,
    String? error,
  }) = _ScanState;

  factory ScanState.fromJson(Map<String, dynamic> json) =>
      _$ScanStateFromJson(json);
}
